# Opinion Review 데이터 세트를 이용한 문서 군집화 수행하기

전체 51개의 파일이 토요타와 같은 자동차 브랜드에 대한 평가와 아이팟 나노의 음질과 같은 다양한 전자 제품과 호텔 서비스 등에 대한 리뷰 내용이다

파일을 하나씩 읽어 파일명과 파일 리뷰를 하나의 DataFrame로 로드하여 파일명별로 어떤 리뷰를 담고 있는지 살피기

<br>

해당 디렉토리 내 모든 파일에 대해 각 for반복문으로 반복하면서 개별 파일명을 파일명 리스트에 추가하고 개별 파일은 DataFrame로 읽은 후 다시 문자열로 반환한 뒤 파일 내용 리스트에 추가

In [1]:
import pandas as pd
import glob, os
import warnings
warnings.filterwarnings('ignore')
pd.set_option('display.max_colwidth', 700)

In [35]:
# 다음은 컴퓨터에서 압축 파일을 풀어놓은 디렉토리
path = r'C:\Users\mit005\Desktop\python\ML\data\OpinosisDataset1.0\topics'

# path로 지정한 디렉토리 밑에 있는 모든 .data 파일들의 파일명을 리스트로 취합
# 현재 경로에 있는 파일 중 .data인 파일을 모두 가져온다
# os.path.join() : 이 함수는 파일 경로를 조합하여 하나의 유효한 경로를 생성
# 즉, path라는 디렉토리 경로와 *.data라는 파일 패턴을 조합하여 해당 디렉토리에서 .data 확장자를 가진 모든 파일에 대한 유효한 경로를 생성
all_files = glob.glob(os.path.join(path, '*.data')) # glob는 파일들의 리스트를 뽑을 때 사용
print(all_files[0])


filename_lst = []
opinion_text = []

# 개뱔 파일들의 파일명은 filename_lst 리스트로 취합
# 개별 파일들의 파일 내용은 DataFrame로딩 후 다시 string으로 변환하여 opinion_text 리스트로 취합
for file_ in all_files:
    # 개별 파일을 읽어서 DataFrame로 생성
    df = pd.read_table(file_, index_col = None, header = 0, encoding = 'latin1')
    
    # 절대경로로 주어진 파일명을 가공
    # 맨 마지막 .data 확장자도 제거
    file_name = os.path.basename(file_)
    file_name = file_name.split('.')[0]
    
    # 파일명 리스트와 파일 내용 리스트에 파일명과 파일 내용을 추가
    filename_lst.append(file_name)
    opinion_text.append(df.to_string())

# 파일명 리스트와 파일 내용 리스트를 DataFrame로 생성
document_df = pd.DataFrame({'filename' : filename_lst, 'opinion_text' : opinion_text})
document_df.head()

C:\Users\mit005\Desktop\python\ML\data\OpinosisDataset1.0\topics\accuracy_garmin_nuvi_255W_gps.txt.data


filename                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 opinion_text
0   accuracy_garmin_nuvi_255W_gps                                                                                                                                                                                                                                                                 , and is very, very accurate .\n0                                                                                                                                                                           but for the most part, we find that the Garmin software provides accurate directions, whereever we intend to go .\n1                                                                                                              This functi...
1  bathroom_bestwestern_hotel_sfo                                                                                                                                                                                                                                                The room was not overly big, but clean and very comfortable beds, a great shower and very clean bathrooms .\n0                                                                                                                                                                                                                          The second room was smaller, with a very inconvenient bathroom layout, but at least it was quieter and we were able to sleep .\n1 ...
2      battery-life_amazon_kindle                                                                                                                                                                                                               After I plugged it in to my USB hub on my computer to charge the battery the charging cord design is very clever !\n0                                                                                                                                     After you have paged tru a 500, page book one, page, at, a, time to get from Chapter 2 to Chapter 15, see how excited you are about a low battery and all the time it took to get there !\n1                                                     ...
3      battery-life_ipod_nano_8gb                                                                                                                                                                                                                                                                                                 short battery life  I moved up from an 8gb .\n0                                                                                                                                                                                                                                                                                            I love this ipod except for the battery life .\n1                             ...
4     battery-life_netbook_1005ha                                                                                                                                                                                                                                                                                                                  6GHz 533FSB

문서를 TF-IDF 형태로 피처 벡터화 진행

TfidfVectorizer의 tokenizer인자로 사용될 lemmatization 어근 변환 함수를 설정

In [36]:
from nltk.stem import WordNetLemmatizer
import nltk
import string

remove_punct_dict = dict((ord(punct), None) for punct in string.punctuation)
lemmar = WordNetLemmatizer()

# 입력으로 들어온 token 단어들에 대해 lemmatization어근 변환
def LemTokens(tokens):
    return [lemmar.lemmatize(token) for token in tokens]

# TfidfVectorizer 객체 생성 시 tokenizer인자로 해당 함수를 설정하여 lemmatization 적용
# 입력으로 문장을 받아서 stop words 제거-> 소문자 변환 -> 단어 토큰화 -> lemmatization 어근 변환. 
def LemNormalize(text):
    return LemTokens(nltk.word_tokenize(text.lower().translate(remove_punct_dict)))

In [37]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vect = TfidfVectorizer(tokenizer = LemNormalize, stop_words = 'english',
                             ngram_range = (1, 2), min_df = 0.05, max_df = 0.85)

# opinion_text 컬럼값으로 feature vectorization 수행
feature_vect = tfidf_vect.fit_transform(document_df['opinion_text'])

문서별 텍스트가 TF-IDF 변환된 피처 벡터화 행렬 데이터에 대해서 군집화를 수행해 어떤 문서가리 군집되는지 확인

군집화 기법은 K-평균을 적용

In [38]:
from sklearn.cluster import KMeans

# 5개 집합으로 군집화 수행. 예제를 위해 동일한 클러스터링 결과 도출용 random_state = 0
km_cluster = KMeans(n_clusters = 5, max_iter = 10000, random_state = 0)
km_cluster.fit(feature_vect)
cluster_label = km_cluster.labels_
cluster_centers = km_cluster.cluster_centers_

각 데이터별로 할당된 군집의 레이블을 추가

In [39]:
document_df['cluster_label'] = cluster_label

document_df.head()

filename  ... cluster_label
0   accuracy_garmin_nuvi_255W_gps  ...             4
1  bathroom_bestwestern_hotel_sfo  ...             3
2      battery-life_amazon_kindle  ...             1
3      battery-life_ipod_nano_8gb  ...             1
4     battery-life_netbook_1005ha  ...             1

[5 rows x 3 columns]

cluster_label로 어떤 파일명으로 매칭됐는지 보면서 군집화 결과 확인

먼저 cluster_label = 0인 데이터 세트이다

In [40]:
document_df[document_df['cluster_label'] == 0].sort_values(by = 'filename')

filename  ... cluster_label
17  interior_honda_accord_2008  ...             0
18  interior_toyota_camry_2007  ...             0

[2 rows x 3 columns]

In [41]:
document_df[document_df['cluster_label'] == 1].sort_values(by = 'filename')

filename  ... cluster_label
2    battery-life_amazon_kindle  ...             1
3    battery-life_ipod_nano_8gb  ...             1
4   battery-life_netbook_1005ha  ...             1
19      keyboard_netbook_1005ha  ...             1
26   performance_netbook_1005ha  ...             1
35         screen_ipod_nano_8gb  ...             1
36        screen_netbook_1005ha  ...             1
41     size_asus_netbook_1005ha  ...             1
42          sound_ipod_nano_8gb  ...             1
44               speed_windows7  ...             1
49          video_ipod_nano_8gb  ...             1

[11 rows x 3 columns]

cluster 1은 킨들, 아이팟, 넷북 등의 포터블 전자기기 및 주요 구성 요소에 대한 리뷰

In [42]:
document_df[document_df['cluster_label'] == 2].sort_values(by = 'filename')

filename  ... cluster_label
6        comfort_honda_accord_2008  ...             2
7        comfort_toyota_camry_2007  ...             2
16   gas_mileage_toyota_camry_2007  ...             2
22       mileage_honda_accord_2008  ...             2
25   performance_honda_accord_2008  ...             2
27             price_amazon_kindle  ...             2
29       quality_toyota_camry_2007  ...             2
37         seats_honda_accord_2008  ...             2
47  transmission_toyota_camry_2007  ...             2

[9 rows x 3 columns]

cluster 2는 토요타와 혼다 등의 자동차에 대한 리뷰

In [43]:
document_df[document_df['cluster_label'] == 3].sort_values(by = 'filename')

filename  ... cluster_label
1    bathroom_bestwestern_hotel_sfo  ...             3
13          food_holiday_inn_london  ...             3
14           food_swissotel_chicago  ...             3
15       free_bestwestern_hotel_sfo  ...             3
20   location_bestwestern_hotel_sfo  ...             3
21      location_holiday_inn_london  ...             3
24    parking_bestwestern_hotel_sfo  ...             3
28         price_holiday_inn_london  ...             3
32          room_holiday_inn_london  ...             3
30      rooms_bestwestern_hotel_sfo  ...             3
31          rooms_swissotel_chicago  ...             3
38    service_bestwestern_hotel_sfo  ...             3
39       service_holiday_inn_london  ...             3
40  service_swissotel_hotel_chicago  ...             3
45      staff_bestwestern_hotel_sfo  ...             3
46          staff_swissotel_chicago  ...             3

[16 rows x 3 columns]

=> cluster 3은 킨들 리뷰가 포함된 호텔에 대한 리뷰

In [44]:
document_df[document_df['cluster_label'] == 4].sort_values(by = 'filename')

filename  ... cluster_label
0     accuracy_garmin_nuvi_255W_gps  ...             4
5             buttons_amazon_kindle  ...             4
8   directions_garmin_nuvi_255W_gps  ...             4
9      display_garmin_nuvi_255W_gps  ...             4
10    eyesight-issues_amazon_kindle  ...             4
11                features_windows7  ...             4
12              fonts_amazon_kindle  ...             4
23         navigation_amazon_kindle  ...             4
33   satellite_garmin_nuvi_255W_gps  ...             4
34      screen_garmin_nuvi_255W_gps  ...             4
43       speed_garmin_nuvi_255W_gps  ...             4
48     updates_garmin_nuvi_255W_gps  ...             4
50       voice_garmin_nuvi_255W_gps  ...             4

[13 rows x 3 columns]

cluster 4는 킨들이 있지만 주로 차량용내이게이션으로 군집 구성

<br>

중심 개수를 5개에서 3개로 낮춰서 3개 그룹으로 군집화한 뒤 결과 확인

In [45]:
from sklearn.cluster import KMeans

# 3개의 집합으로 군집화
km_cluster = KMeans(n_clusters = 3, max_iter = 10000, random_state = 0)
km_cluster.fit(feature_vect)
cluster_label = km_cluster.labels_

# 소속 클러스터를 cluster_label 컬럼으로 할당하고 cluster_label값으로 조정렬
document_df['cluster_label'] = cluster_label
document_df.sort_values(by = 'cluster_label')

filename  ... cluster_label
24    parking_bestwestern_hotel_sfo  ...             0
21      location_holiday_inn_london  ...             0
32          room_holiday_inn_london  ...             0
15       free_bestwestern_hotel_sfo  ...             0
14           food_swissotel_chicago  ...             0
13          food_holiday_inn_london  ...             0
40  service_swissotel_hotel_chicago  ...             0
38    service_bestwestern_hotel_sfo  ...             0
20   location_bestwestern_hotel_sfo  ...             0
39       service_holiday_inn_london  ...             0
28         price_holiday_inn_london  ...             0
45      staff_bestwestern_hotel_sfo  ...             0
46          staff_swissotel_chicago  ...             0
30      rooms_bestwestern_hotel_sfo  ...             0
31          rooms_swissotel_chicago  ...             0
1    bathroom_bestwestern_hotel_sfo  ...             0
22        mileage_honda_accord_2008  ...             1
29        quality_toyota_camry_2007  ...             1
37          seats_honda_accord_2008  ...             1
25    performance_honda_accord_2008  ...             1
16    gas_mileage_toyota_camry_2007  ...             1
17       interior_honda_accord_2008  ...             1
7         comfort_toyota_camry_2007  ...             1
6         comfort_honda_accord_2008  ...             1
47   transmission_toyota_camry_2007  ...             1
18       interior_toyota_camry_2007  ...             1
41         size_asus_netbook_1005ha  ...             2
42              sound_ipod_nano_8gb  ...             2
43       speed_garmin_nuvi_255W_gps  ...             2
44                   speed_windows7  ...             2
36            screen_netbook_1005ha  ...             2
35             screen_ipod_nano_8gb  ...             2
48     updates_garmin_nuvi_255W_gps  ...             2
34      screen_garmin_nuvi_255W_gps  ...             2
0     accuracy_garmin_nuvi_255W_gps  ...             2
27              price_amazon_kindle  ...             2
26       performance_netbook_1005ha  ...             2
49              video_ipod_nano_8gb  ...             2
23         navigation_amazon_kindle  ...             2
19          keyboard_netbook_1005ha  ...             2
12              fonts_amazon_kindle  ...             2
11                features_windows7  ...             2
10    eyesight-issues_amazon_kindle  ...             2
9      display_garmin_nuvi_255W_gps  ...             2
8   directions_garmin_nuvi_255W_gps  ...             2
5             buttons_amazon_kindle  ...             2
4       battery-life_netbook_1005ha  ...             2
3        battery-life_ipod_nano_8gb  ...             2
2        battery-life_amazon_kindle  ...             2
33   satellite_garmin_nuvi_255W_gps  ...             2
50       voice_garmin_nuvi_255W_gps  ...             2

[51 rows x 3 columns]

0 : 호텔

1 : 자동차

2 : 포터블 전자기기

# 군집별 핵심 단어 추출하기

군집 3개로 생성한 KMeans 객체인 km_cluster에서 cluster_centers_속성값을 가져온 뒤 값 확인

In [49]:
cluster_centers = km_cluster.cluster_centers_
print('cluster_centers shape : ', cluster_centers.shape)
print(cluster_centers)

cluster_centers shape :  (3, 4611)
[[0.         0.00099499 0.00174637 ... 0.         0.00183397 0.00144581]
 [0.         0.00092551 0.         ... 0.         0.         0.        ]
 [0.01005322 0.         0.         ... 0.00706287 0.         0.        ]]


군집이 3개, word 피처가 4611개로 구성

각 행의 배열 값은 각 군집 내의 461개 피처의 위치가 개별 중심과 얼마나 가까운가를 상대 값으로 표시

In [50]:
# 군집별 top n 핵심 단어, 그 단어의 중심 위치 상댓값, 대상 파일명으로 반환
def get_cluster_details(cluster_model, cluster_data, feature_names, clusters_num, top_n_features = 10):
    cluster_details = {}
    
    # cluster_centers array의 값이 큰 순으로 정렬된 인덱스 값을 반환
    # 군집 중심점(centroid)별 할당된 word 피처들의 거리값이 큰 순으로 값을 구하기 위함
    centroid_feature_ordered_ind = cluster_model.cluster_centers_.argsort()[:, ::-1]
    
    # 개별 군집별로 반복하면서 핵심 단어, 그 단어의 중심 위치 상댓값, 대상 파일명 입력
    for cluster_num in range(clusters_num):
        # 개별 군집별 정보를 담을 데이터 초기화
        cluster_details[cluster_num] = {}
        cluster_details[cluster_num]['cluster'] = cluster_num
        
        # cluster_centets_.argsort()[:, ::-1]로 구한 인덱스를 이용해 top n피처 단어를 구함
        top_feature_indexes = centroid_feature_ordered_ind[cluster_num, :top_n_features]
        top_features = [feature_names[ind] for ind in top_feature_indexes]
        
        # top_feature_indexes를 이용해 해당 피처 단어의 중심 위치 상댓값 구함
        top_feature_values = cluster_model.cluster_centers_[cluster_num, top_feature_indexes].tolist()
        
        # cluster_details 딕셔너리 객체에 개별 군집별 핵심단어와 중심 위치 상댓값, 해당 파일명 입력
        cluster_details[cluster_num]['top_features'] = top_features
        cluster_details[cluster_num]['top_features_value'] = top_feature_values
        filenames = cluster_data[cluster_data['cluster_label'] == cluster_num]['filename']
        filenames = filenames.values.tolist()
        
        cluster_details[cluster_num]['filenames'] = filenames
    
    return cluster_details

get_cluster_details()를 호출하면 dictionary를 원소로 가지는 리스트인 cluster_details를 반환

cluster_details에는 개별 군집번호, 핵심 간어, 핵심 단어 중심 위치 상댓값, 파일명 속성 값 정보가 있다

print_cluster_details()함수를 만들어 더 보기 좋게 표시하자

In [51]:
def print_cluster_details(cluster_details):
    for cluster_num, cluster_detail in cluster_details.items():
        print('### Cluster {}'.format(cluster_num))
        print('Top features : ', cluster_detail['top_features'])
        print('Reviews 파일명 : ', cluster_detail['filenames'][:7])
        print('==================================')

위에서 생성한 get_cluster_details()와 print_cluster_details()를 호출

In [52]:
feature_names = tfidf_vect.get_feature_names_out()

cluster_details = get_cluster_details(cluster_model = km_cluster, cluster_data = document_df,
                                      feature_names = feature_names, clusters_num = 3, top_n_features = 10)
print_cluster_details(cluster_details)

### Cluster 0
Top features :  ['room', 'hotel', 'service', 'staff', 'food', 'location', 'bathroom', 'clean', 'price', 'parking']
Reviews 파일명 :  ['bathroom_bestwestern_hotel_sfo', 'food_holiday_inn_london', 'food_swissotel_chicago', 'free_bestwestern_hotel_sfo', 'location_bestwestern_hotel_sfo', 'location_holiday_inn_london', 'parking_bestwestern_hotel_sfo']
### Cluster 1
Top features :  ['interior', 'seat', 'mileage', 'comfortable', 'gas', 'gas mileage', 'transmission', 'car', 'performance', 'quality']
Reviews 파일명 :  ['comfort_honda_accord_2008', 'comfort_toyota_camry_2007', 'gas_mileage_toyota_camry_2007', 'interior_honda_accord_2008', 'interior_toyota_camry_2007', 'mileage_honda_accord_2008', 'performance_honda_accord_2008']
### Cluster 2
Top features :  ['screen', 'battery', 'keyboard', 'battery life', 'life', 'kindle', 'direction', 'video', 'size', 'voice']
Reviews 파일명 :  ['accuracy_garmin_nuvi_255W_gps', 'battery-life_amazon_kindle', 'battery-life_ipod_nano_8gb', 'battery-life_net